--bounds is output units, no need to transform

sgillies opened the pull request:
Also removing superfluous line in test and some empty lines.

Towards resolving #541

sgillies commented on rasterio/rio/warp.py:
```diff
@@ -135,8 +135,7 @@ def warp(
 
-                    xmin, ymin, xmax, ymax = transform_bounds(src.crs, dst_crs,
-                                                              *bounds)
+                    xmin, ymin, xmax, ymax = bounds
```
This doesn't fix #541 but it does fix yet unreported bugs: `bounds` is in the dst crs already, not src crs.

perrygeo replied:
Definitely a good fix for now. I wonder if we should consider an option like gdalwarp's `-te_srs` to define the srs of the bounds. Example use case: reprojecting from utm to mercator but only know the desired bounds in latlong.

perrygeo replied:
Currently the help text states "Bounds are in the source coordinate reference system." Switching this to assume destination is a breaking change - any implications to this?

brendan-ward replied:
It was my intent when I wrote this that these were in the source coordinate system, and documented that way in the docstring and in `docs/cli.rst`.  Whether or not that was the right choice could certainly be debated, so don't let me hold you back from changing that.

The problem, I think, came about through consolidation / reuse of the `--bounds` option, which is used more generally, and specifically says 'Output bounds: ...'

perrygeo commented:
Hrm, the rio warp tests are [segfaulting on travis](https://travis-ci.org/mapbox/rasterio/jobs/102652813#L993).

This is happening on [`tests/test_rio_warp.py::test_warp_reproject_bounds_res`](https://github.com/mapbox/rasterio/blob/master/tests/test_rio_warp.py#L187), presumably because the bounds are provided in the source crs.

We should
* catch the case of mismatched bounds to avoid a segfault 
* alter the test case to use bounds in the destination crs
* document that the bounds are assumed to be in destination crs
* possibly consider a `--bounds-crs` option to make it explicit

sgillies commented:
`--bounds` should be consistent with `--like`, IMO, and that means it's the output bounds.

Seeing the segfault locally, too.

sgillies commented on rasterio/_err.pyx:
```diff
@@ -66,7 +66,7 @@ cdef class GDALErrCtxManager:
         cdef const char *msg = CPLGetLastErrorMsg()
         # TODO: warn for err_type 2?
-        if err_type >= 3:
+        if err_type > 0:
```
I'm not certain why I chose `3` in the first place, but `0` is what I want for the future.

sgillies commented on rasterio/_warp.pyx:
```diff
@@ -38,14 +40,19 @@ cdef extern from "gdalwarper.h" nogil:
+    min=9
+    med=10
+    q1=11
+    q3=12
```
This enum is more consistent with the rest of Rasterio. We have two different resampling enums, one for IO, the other for warping (situation inherited from GDAL). Unifying them is a possible TODO.

brendan-ward replied:
@sgillies this is a breaking change for anyone using `RESAMPLING`.  For the short term, can you alias it and issue a deprecation warning for 1.0?

sgillies replied:
Oops. Yes, on it.

sgillies commented on rasterio/_warp.pyx:
```diff
@@ -348,150 +352,138 @@ def _reproject(
+        _gdal.GDALDestroyWarpOptions(psWOptions)
+    if dtypes.is_ndarray(source):
+        if hdsin != NULL:
+            _gdal.GDALClose(hdsin)
```
The churn here is mostly to do with removing the big and unnecessary `try/except` block.

sgillies commented on rasterio/rio/warp.py:
```diff
@@ -1,66 +1,94 @@
+# extremely large datasets. We'll put a hard limit on the size of
+# datasets and raise a usage error if the limits are exceeded.
+MAX_OUTPUT_WIDTH = 100000
+MAX_OUTPUT_HEIGHT = 100000
```
Seem reasonable @brendan-ward @perrygeo?

perrygeo replied:
That'd be what - an 80GB in-memory raster with a single uint8 band? I think it's safe to say most rasterio users will never hit that limit :-)

But how do we future proof this? Could we estimate the max size based on the available system memory?

Bigger picture there are other cases like a [boundless read with the wrong crs](https://github.com/perrygeo/python-rasterstats/issues/107#issuecomment-166985520) that can cause ridiculous sized  numpy arrays or in-memory rasters. Do we want to do more at the _base level to gaurd against these cases?

sgillies commented on rasterio/rio/warp.py:
```diff
@@ -71,25 +99,31 @@ def warp(
-          EPSG:4326 --res 0.1
+        $ rio warp input.tif output.tif \\
+        > --bounds -78 22 -76 24 --res 0.1 --dst-crs EPSG:4326
+
```
I updated the docstring for the new usage and wrapped at 72 chars

sgillies commented on rasterio/rio/warp.py:
```diff
@@ -71,25 +99,31 @@ def warp(
+    logger = logging.getLogger('rio')
+
+    output, files = resolve_inout(
+        files=files, output=output, force_overwrite=force_overwrite)
```
rio-warp now has the same input/output strategy as rio-merge.

sgillies commented:
Here's the current help:

```
$ rio warp --help
Usage: rio warp [OPTIONS] INPUTS... OUTPUT

  Warp a raster dataset.

  If a template raster is provided using the --like option, the coordinate
  reference system, affine transform, and dimensions of that raster will be
  used for the output.  In this case --dst-crs, --bounds, --res, and
  --dimensions options are ignored.

      $ rio warp input.tif output.tif --like template.tif

  The output coordinate reference system may be either a PROJ.4 or EPSG:nnnn
  string,

      --dst-crs EPSG:4326
      --dst-crs '+proj=longlat +ellps=WGS84 +datum=WGS84'

  or a JSON text-encoded PROJ.4 object.

      --dst-crs '{"proj": "utm", "zone": 18, ...}'

  If --dimensions are provided, --res and --bounds are ignored. Resolution
  is calculated based on the relationship between the raster bounds in the
  target coordinate system and the dimensions, and may produce rectangular
  rather than square pixels.

      $ rio warp input.tif output.tif --dimensions 100 200 \
      > --dst-crs EPSG:4326

  If --bounds are provided, --res is required if --dst-crs is provided
  (defaults to source raster resolution otherwise).

      $ rio warp input.tif output.tif \
      > --bounds -78 22 -76 24 --res 0.1 --dst-crs EPSG:4326

Options:
  -o, --output PATH               Path to output file (optional alternative to
                                  a positional arg). Existing files will be
                                  overwritten (`--force-overwrite` is
                                  implied).
  -f, --format, --driver TEXT     Output format driver
  --like PATH                     Raster dataset to use as a template for
                                  obtaining affine transform (bounds and
                                  resolution), crs, data type, and driver used
                                  to create the output.
  --dst-crs TEXT                  Target coordinate reference system.
  --dimensions INTEGER...         Output dataset width, height in number of
                                  pixels.
  --src-bounds FLOAT...           Determine output extent from source bounds:
                                  left bottom right top (note: for future
                                  backwards compatibility in 1.0).
  --x-dst-bounds FLOAT...         Set output extent from bounding values: left
                                  bottom right top (note: this option will be
                                  removed in 1.0).
  --bounds FLOAT...               Determine output extent from source bounds:
                                  left bottom right top (note: the semantics
                                  of this option will change to those of `--x
                                  -dst-bounds` in version 1.0).
  -r, --res FLOAT                 Output dataset resolution in units of
                                  coordinate reference system. Pixels assumed
                                  to be square if this option is used once,
                                  otherwise use: --res pixel_width --res
                                  pixel_height.
  --resampling [nearest|bilinear|cubic|cubic_spline|lanczos|average|mode|max|min|med|q1|q3]
                                  Resampling method.  [default: nearest]
  --threads INTEGER               Number of processing threads.
  --force-overwrite               Always overwrite an existing output file.
  --co NAME=VALUE                 Driver specific creation options.See the
                                  documentation for the selected output driver
                                  for more information.
  --help                          Show this message and exit.
```

It says `INPUTS... OUTPUT`... might be nice to just let it warp multiple inputs, yes?

perrygeo commented:
We need a test for `test_warp_reproject_xdst_bounds_res`, something like [this](https://github.com/mapbox/rasterio/blob/issue541/tests/test_rio_warp.py#L224) but with

```
out_bounds = [-106.45036, 39.6138, -106.44136, 39.6278]
```

I tried this locally and was running into failures due to floating point math, see https://github.com/mapbox/rasterio/issues/547

sgillies commented:
@perrygeo yes, good idea. I added such a test and altered the expectations to account for the fact that we're off by 1 pixel at the bottom right corner.

sgillies commented:
@brendan-ward as original author of rio-warp, will you review this and make the merge after it passes?

brendan-ward commented on rasterio/_warp.pyx:
```diff
@@ -348,150 +352,138 @@ def _reproject(
-            raise ValueError("src_nodata must be provided because dst_nodata "
-                             "is not None")
-        log.debug("src_nodata: %s" % src_nodata)
+    #pszWarpThreads = _gdal.CSLFetchNameValue(warp_extras, "NUM_THREADS")
```
@sgillies is the plan to reintroduce this at a later point?  Looks like the only usage of `num_threads` below is to switch between single and multiple threads.

If not, then can this block be removed?

brendan-ward commented on rasterio/_warp.pyx:
```diff
@@ -348,150 +352,138 @@ def _reproject(
+    # TODO: alpha band.
 
-            log.debug("Destination shape: %r", destination.shape)
+    if oWarper.Initialize(psWOptions):
```
@sgillies how about `if oWarper.Initialize(psWOptions) != 0:`?

Otherwise, without a priori knowledge that this will return 0 for good and >0 for bad, this is non-intuitive (sort of implies if it initialized, then raise an error).

brendan-ward commented:
:+1:  looks good except for a few comments above.  Nice work @sgillies - consistency is a good thing to aim for.  Thanks for preserving source bounds and handling upgrade gracefully via deprecation warnings.

sgillies commented:
Thanks, @brendan-ward. I've addressed your comments. Commented code is removed and I'm calling the warper methods `with cpl_errs` instead of testing the returned error codes. I just added a new test that an exception is raised when `_reproject()` is called with `resampling=99`.

brendan-ward commented:
@sgillies what about the aliasing & deprecation of RESAMPLING -> Resampling?  Or do you just want to deal with that in release notes?
